src/coreclr/System.Private.CoreLib/src/System/Attribute.CoreCLR.cs

```diff
@@ -438,6 +438,9 @@ private static Attribute[] CreateAttributeArrayHelper(Type elementType, int elem
         {
             return (Attribute[])Array.CreateInstance(elementType, elementCount);
         }
+
+        private static Attribute[] ToAttributeArray(object[] getCustomAttributesResult) =>
```

Comment:
Is there an actual real situation where `Array.Empty<Attribute>();` gets exercised now?

All places where this is called specify that the attribute has to inherit from `Attribute` or check that the requested attribute type inherits from `Attribute`. It means that we will never end up with an array that is not assignable to `Attribute[]` here, unless there is a bug. I think replacing the `!` with throwing cast would work just fine, and it would be simpler than this fix.

Reply:
> replacing the ! with throwing cast would work just fine, and it would be simpler than this fix.

I've done this.

> Is there an actual real situation where Array.Empty<Attribute>(); gets exercised now?

Not sure whether this is realistic or a scenario we care about, but there is a situation that can hit this path. Because anyone can create custom `MemberInfo` types (e. g. by deriving from `FieldInfo`, they can create a type whose `GetCustomAttributes()` function returns null or a non-`Attribute` array type.

I added a test case that shows this path.

Reply:
Good point. I do not think we care about rare cases where the API violate their contracts. (A similar discussion comes up e.g. for Streams that violate the stream contract.)